### PR TITLE
Cogljj/lazy import

### LINKED ISSUE #1182: [TASK] Make tensorflow loaded lazily
--------
Issue Description
--------
Right now tensorflow is always loaded, even if the input does not require it. This slows down every run of RAVEN (from a few seconds to 30 seconds depending on the computer).  If it could be loaded only on demand, it would still be available for the inputs that use it, but not for the other RAVEN runs.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by joshua-cogliati-inl**
--------
Pull Request Description
--------
##### What issue does this change request address? 
Closes #1182 

##### What are the significant changes in functionality due to this change request?
This adds the TensorFlow lazy loading library, and removes lazy-import. It uses the lazy loading library to load TensorFlow and switches the other lazy loading to it.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: framework/SupervisedLearning/KerasClassifier.py [hunk 1]
```diff
@@ -32,21 +32,24 @@
 from sklearn import preprocessing
 import os
 
+import contrib.lazy.lazy_loader as lazy_loader
```
Comment (alfoa) [inline framework/SupervisedLearning/KerasClassifier.py]: josh we created in utils an lazyImporterUtils.py....maybe you want to modify it

Comment (joshua-cogliati-inl) [reply framework/SupervisedLearning/KerasClassifier.py]: Thanks. (Also, is there a good method of testing the loading time of Raven?)

Comment (alfoa) [reply framework/SupervisedLearning/KerasClassifier.py]: "generally" I test it when the "PRA" plugin gets loaded.

Comment (alfoa) [reply framework/SupervisedLearning/KerasClassifier.py]: I know it is not ideal ;/

Comment (joshua-cogliati-inl) [reply framework/SupervisedLearning/KerasClassifier.py]: Looks like I need to move the availOptimizers and other things to the ```__init__```

**Comment by joshua-cogliati-inl** [comment]
Looks like I need to move the availOptimizers and other things to the ```__init__``` in order to get lazy loading working.

File: framework/utils/lazyImporterUtils.py [hunk 2]
```diff
@@ -34,13 +34,14 @@ def import_module_lazy(moduleString):
...
-  """
-  return lazy_import.lazy_callable(collableString.strip())
+#def import_collable_lazy(collableString):
+#  """
+#    This method is aimed to import a collable method or attribute
+#    within a module that needs to be lazy imported
+#    @ In, moduleString, str, the collable to point (e.g. numpy.arange, numpy.ndarray, etc.)
+#    @ Out, callable, Object_Pointer, the collable (lazy)
+#  """
+#  return lazy_import.lazy_callable(collableString.strip())
```
Comment (alfoa) [inline framework/utils/lazyImporterUtils.py]: we can remove it

**Comment by alfoa** [comment]
can you create an issue about this for tracking purpose (and close it in the pr description)

File: framework/utils/lazyImporterUtils.py (hunk 2 above)
Comment (joshua-cogliati-inl) [reply framework/utils/lazyImporterUtils.py]: Removed.

File: framework/SupervisedLearning/KerasClassifier.py (hunk 1 above)
Comment (joshua-cogliati-inl) [reply framework/SupervisedLearning/KerasClassifier.py]: I added -v to the debug python command in raven_framework and that can be used to check which modules are being loaded.

Comment (alfoa) [reply framework/SupervisedLearning/KerasClassifier.py]: top

Comment (alfoa) [reply framework/SupervisedLearning/KerasClassifier.py]: did you push the modified script?

Comment (joshua-cogliati-inl) [reply framework/SupervisedLearning/KerasClassifier.py]: I did not push the modified raven_framework script.

Comment (alfoa) [reply framework/SupervisedLearning/KerasClassifier.py]: ah okay...maybe we can write it in the developer wiki ? just for reference?

**Comment by alfoa** [comment]
Checklist passed...

PR can be accepted...

**Pull request merged**